### ISSUE TITLE
macOS kernel panic during boot due to non-monotonic TSC

### ISSUE THREAD

**Issue by sej7278** [thenickdude/KVM-Opencore#15]
Hi, this is possibly not even an opencore issue, but does anything jump out at you as being an obvious problem?

Got a VM running 11.6.1 with your oc 0.7.4 install on my Xeon E5-2650v2, but it simply will not take the Monterey update (also didn't do one of the beta's i tried or a fresh install) it starts the upgrade but after a couple of reboots seems to panic and get stuck at mach reboot.

It gets past the time-out here:

![1_timeout](https://user-images.githubusercontent.com/817564/138937179-f8f9760a-21da-40fc-b5bf-58719d005b15.jpg)

But then panics here:

![2_panic](https://user-images.githubusercontent.com/817564/138937236-592c145b-eb84-4774-ac39-aecce3628698.jpg)

And finally hangs completely here trying to reboot i guess:

![3_mach_reboot](https://user-images.githubusercontent.com/817564/138937286-68987993-c421-4635-8bb0-bd04c5d4c953.jpg)

My config.plist is basically the same as yours but with some smbios stuff added by opencore-configurator (serials etc.) oc-validator had nothing bad to say about it.

libvirt xml is: [libvirt.txt](https://github.com/thenickdude/KVM-Opencore/files/7420648/libvirt.txt)

Host is Debian Sid, qemu 6.1.0, libvirt 7.6.0, kernel 5.14.12

**Comment by sej7278** [comment]
@sickcodes yes it's definitely doing that but I think I'm making it past that stage

**Comment by sej7278** [comment]
i think i finally managed to get a screenshot before it crashes, if this makes any sense:

![Screenshot from 2021-10-30 00-41-40](https://user-images.githubusercontent.com/817564/139511938-4de16932-d16a-410c-aa9f-1b6b855ca56b.png)

**Comment by thenickdude** [comment]
So that seems to be panicing due to non-monotonic time (clock going backwards).

I wonder if you're getting a warning at VM launch time that "invtsc" isn't actually available on your system. Try removing that from your CPU args if it's currently there.

Can you post the VM command/config you're currently using and also the output of this on the host:

    cat /proc/cpuinfo

(You only need to paste the output from a single one of the cores)

**Comment by thenickdude** [comment]
This user has the same panic on bare-metal:

https://www.reddit.com/r/hackintosh/comments/qhjnly/random_kernel_panics_on_x79/

**Comment by thenickdude** [comment]
Also, your host isn't going to sleep during the install because you aren't moving the mouse to keep it awake, is it?

**Comment by sej7278** [comment]
ah i had a problem with kvm-pit with catalina hard crashing the host, the fix was to remove this lot, but i don't have that in my monterey config, i wonder what the defaults are:

```xml
<clock offset='utc'>
    <timer name='rtc' tickpolicy='catchup'/>
    <timer name='pit' tickpolicy='delay'/>
    <timer name='hpet' present='no'/>
</clock>
```

cpuinfo:

```
processor	: 31
vendor_id	: GenuineIntel
cpu family	: 6
model		: 62
model name	: Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz
stepping	: 4
microcode	: 0x42e
cpu MHz		: 1200.000
cache size	: 20480 KB
physical id	: 1
siblings	: 16
core id		: 7
cpu cores	: 8
apicid		: 47
initial apicid	: 47
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm cpuid_fault epb pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase smep erms xsaveopt dtherm ida arat pln pts md_clear flush_l1d
vmx flags	: vnmi preemption_timer posted_intr invvpid ept_x_only ept_1gb flexpriority apicv tsc_offset vtpr mtf vapic ept vpid unrestricted_guest vapic_reg vid ple
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit
bogomips	: 5190.81
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

host is a desktop so not going to sleep.

when i got that screenshot i was running OpenCore-boot.sh instead of virt-manager.

i do get this in dmesg, don't know how to fix it though:

```
dmesg |grep -i kvm
[  215.019676] kvm: SMP vm created on host with unstable TSC; guest TSC will not be reliable

cat /sys/devices/system/clocksource/clocksource0/available_clocksource 
hpet acpi_pm 

cat /sys/devices/system/clocksource/clocksource0/current_clocksource 
hpet
```

removing +invtsc isn't making any difference

**Comment by thenickdude** [comment]
Okay, I'm pretty sure that's your problem. On my system the clocksource is set to tsc, but on yours it doesn't even get offered as an option.

What does this return "dmesg | grep -i tsc"?

I *think* on my system there was an option buried in the host UEFI settings for TSC synchronisation between sockets. If yours has that too, make sure it's turned on, because otherwise it might cause the TSC to be rejected.

I guess since you only have two clocksources to choose from you could try switching to the other and see if things improve:

    echo acpi_pm > /sys/devices/system/clocksource/clocksource0/current_clocksource

**Comment by thenickdude** [comment]
Also it sounds like recent Linux kernels 5.13, 5.14 have made timing changes that cause it to disable TSC in more situations:

https://www.phoronix.com/forums/forum/software/general-linux-open-source/1283799-linux-5-15-rc5-x86-changes-aim-to-fix-yet-another-hardware-trainwreck

If you can try 5.12 and see if the tsc clocksource comes back that would be interesting.

**Comment by sej7278** [comment]
i'll have a look in my bios tomorrow (don't a massive backup right now) but i'm in bios mode not uefi and it is the latest (dell t5610 doesn't get updated very often!). i do recall some time settings but think it was just utc.

```
# dmesg | grep -i -e tsc -e clocksource
[    0.000000] tsc: Fast TSC calibration using PIT
[    0.000000] tsc: Detected 2593.895 MHz processor
[    0.023471] TSC deadline timer available
[    0.023542] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645519600211568 ns
[    0.097315] clocksource: hpet: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 133484882848 ns
[    0.117335] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x2563b4fbe6b, max_idle_ns: 440795330438 ns
[    0.149339] TSC synchronization [CPU#0 -> CPU#1]:
[    0.149339] Measured 1430554 cycles TSC warp between CPUs, turning off TSC clock.
[    0.149339] tsc: Marking TSC unstable due to check_tsc_sync_source failed
[    0.357884] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.693344] clocksource: Switched to clocksource hpet
[    0.711473] clocksource: acpi_pm: mask: 0xffffff max_cycles: 0xffffff, max_idle_ns: 2085701024 ns
[  215.019676] kvm: SMP vm created on host with unstable TSC; guest TSC will not be reliable
```

i've had this TSC issue since 5.10 kernel as i recall, its not new to 5.14

**Comment by thenickdude** [comment]
Yeah if the TSC only differs between sockets then pinning to a single socket should solve this.

It sounds like the sockets need their RESET signal to be delivered at the same time to have in-sync TSCs. It also sounds like the BIOS can ruin the sync by attempting to set the TSC register.

I noticed there is an Intel errata for the whole E5 v2 line that says the TSC won't be reset by a warm reboot. If this is true then if your TSCs ever go out of sync they would stay out of sync until you perform a cold boot.

**Comment by thenickdude** [comment]
This kernel bug report has a patch attached to try to better sync up the TSCs in this case, if firmware updates are not available to fix the core issue:

https://bugzilla.kernel.org/show_bug.cgi?id=202525

**Comment by sej7278** [comment]
cpu pinning made some sort of difference (different crash points?). will try looking into tsc/kernel next.

**Comment by Raikerian** [comment]
I am also having troubles with Monterey on my VM, albeit the issue seems to be different:

![image](https://user-images.githubusercontent.com/8850270/145052758-8a7a6a47-e399-4e24-b142-eb9d1eb8a2b2.png)

Using QEMU 6.0.0 with this options:
```
  -enable-kvm
  -m 16G
  -cpu host,kvm=on,vendor=GenuineIntel,+kvm_pv_unhalt,+kvm_pv_eoi,+hypervisor,+invtsc
  -machine q35,accel=kvm
  -usb -device usb-kbd -device usb-tablet
  -smp 12,cores=2,sockets=3,threads=2,maxcpus=12
  -device usb-ehci,id=ehci
  -device nec-usb-xhci,id=xhci
  -global nec-usb-xhci.msi=off
  -device isa-applesmc,osk=..
  -drive if=pflash,format=raw,readonly=on,file=..
  -drive if=pflash,format=raw,file=..
  -smbios type=2
  -device ich9-intel-hda -device hda-duplex
  -device ich9-ahci,id=sata
  -device ide-hd,bus=sata.4,drive=MacHDD
  -drive id=MacHDD,if=none,file=..,format=qcow2
  -netdev "user,id=net0,hostfwd=tcp::443-:443,hostfwd=tcp::80-:80,hostfwd=tcp::22-:22"
  -device ..,netdev=net0,id=net0,mac=..
  -monitor stdio
  -device VGA,vgamem_mb=128
  -vnc ":0" -k "en-us"
```

Using CPU passthrough here, however I have also tried emulating Penryn and Skylake, but it had 0 effect.

Most of the times it will come to the kernel panic above (I had to turn on full debugging to get that message, otherwise it was never shown, macos would usually reboot around SMC step). But sometimes it actually even boots macos, but then crases almost immediately.
Interesting enough I only have this problem when trying to boot on Intel 3.0 GHz Core i7-4578U, its working perfectly fine on Intel 2.6 GHz Core i5-4278U. However, the i7 machine is booted into linux from external drive connected through USB, which leads me to believe that it is potentially the issue here. Tried playing around with `npci=0x2000` boot flag and setting `ReleaseUsbOwnership` to true, didn't help.

Big Sur is working perfectly fine on both with the same OC and boot options however.

My OVMFs are from OSX-KVM. As for configs, tried different ones already including yours and OSX-KVM ones.
Appreciate any help. Perhaps someone can point me in the direction I can look more into myself as well.

**Comment by Raikerian** [comment]
@thenickdude @sej7278 seems like my problem was also related to clocksource and I was able to solve it, so will leave my notes here, perhaps it will help someone.

With MacOS Monterey and Intel 3.0 GHz Core i7-4578U after many runs most of the times it would crash with the error from the previous message, which is:
```
** In Memory Panic Stackshot Succeeded ** Bytes Traced 15707 (Uncompressed 48592) **
IOPolledInterface::startIO[0] 0xe00002c7
IOPolledFileWrite(0x0xffffff8546c5f480, 0x0, 0, 0x0) : IOStartPolledIO(0x0xffffff8546c5f480, kIOPolledWrite, 0, 0x24c806000, 8192) returned 0xe00002c7
IOPolledFileWrite(gIOPolledCoreFileVars, 0, 0x0, NULL) returned 0xe00002c7
progress_notify_stage_outproc (during forwarding) returned 0xe00002c7
```

Sometimes it would actually get stuck around SMC step, and even rarer it would fully boot into macos, but then almost immediately go into reboot.
On one of the runs I have seen [this error related to clock](https://i.imgur.com/jHSBmRY.png) before the IOPolledFileWrite panic. It was seen pretty rare, and from later findings it seems this one and IOPolledFileWrite panic was actually coming together (but not always displayed due to how fast log updates and I didn't find a way to extract boot log into file as I was not able to boot into macos completely) and are related to the issue linked above https://github.com/acidanthera/bugtracker/issues/1676.

So when I digged into the boot log on the host I have found similar problem to what you guys talked above:
```
$ dmesg | grep -i -e tsc -e clocksource
[    0.000000] tsc: Detected 3200.000 MHz processor
[    0.000000] tsc: Detected 3199.980 MHz TSC
[    0.071145] TSC deadline timer available
[    0.071160] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 1910969940391419 ns
[    0.589882] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x2e2036ff8d5, max_idle_ns: 440795275316 ns
[    0.625855] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 1911260446275000 ns
[    1.147854] clocksource: Switched to clocksource tsc-early
[    1.216141] clocksource: acpi_pm: mask: 0xffffff max_cycles: 0xffffff, max_idle_ns: 2085701024 ns
[    2.928837] tsc: Refined TSC clocksource calibration: 3191.998 MHz
[    2.931728] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x2e02c314322, max_idle_ns: 440795292111 ns
[    2.936036] clocksource: Switched to clocksource tsc
[   14.563926] clocksource: timekeeping watchdog on CPU7: Marking clocksource 'tsc' as unstable because the skew is too large:
[   14.568574] clocksource:                       'acpi_pm' wd_now: d82b24 wd_last: e42ddf mask: ffffff
[   14.572515] clocksource:                       'tsc' cs_now: 33b5415888 cs_last: 30635be285 mask: ffffffffffffffff
[   14.576891] tsc: Marking TSC unstable due to clocksource watchdog
[   14.579764] TSC found unstable after boot, most likely due to broken BIOS. Use 'tsc=unstable'.
[   14.590465] clocksource: Switched to clocksource acpi_pm
[  272.392417] kvm: SMP vm created on host with unstable TSC; guest TSC will not be reliable

$ cat /sys/devices/system/clocksource/clocksource0/current_clocksource
acpi_pm
```

While on the working hardware (Intel 2.6 GHz Core i5-4278U) clocksource would always be tsc. From digging out I found out that in newer versions of kernel [watchdog disabler was added during boot](https://github.com/torvalds/linux/blob/c741e49150dbb0c0aebe234389f4aa8b47958fa8/arch/x86/kernel/tsc.c#L1206-L1221) in case cpu supports tsc features (constant_tsc, nonstop_tsc and tsc_adjust features to be precise) and has 2 or less sockets, which in my case works for my cpu. So adding `tsc=reliable` to the boot options in grub seem to be the temporary workaround here which I am testing. In that case tsc works fine and the host doesn't switch to acpi_pm anymore.

This solved the problem with hanging completely and now I was getting IOPolledFileWrite panic with clock sync problem 100% of the time. So this is where a thread linked above came to help, as it also has a great explanation on why it happens on Monterey: https://github.com/acidanthera/bugtracker/issues/1676#issuecomment-881884751
Just adding new CpuTscSync kext didn't help, so I also updated Lilu kext to the latest version and put CpuTscSync higher on the list in plist file (not sure if this one is necessary though). Finally Monterey boots correctly and I don't have any issues so far anymore.

This was only needed to boot Monterey for me, as BigSur was doing fine with both acpi_pm and no cpu sync kext.

**Comment by thenickdude** [comment]
Thanks for those details!

**Comment by sej7278** [comment]
looks like my bios is buggy, as i tried the `tsc=reliable` kernel param and it booted as tsc then switched back to hpet!

```
# dmesg | grep -i -e tsc -e clocksource
[    0.000000] Command line: BOOT_IMAGE=/vmlinuz-5.15.0-2-amd64 root=UUID=2b82e6c4-442e-4700-a246-3b5c37a722b5 ro tsc=reliable quiet splash slab_common.usercopy_fallback=y intel_iommu=on iommu=pt
[    0.000000] tsc: Fast TSC calibration using PIT
[    0.000000] tsc: Detected 2593.597 MHz processor
[    0.022989] TSC deadline timer available
[    0.023056] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645519600211568 ns
[    0.030289] Kernel command line: BOOT_IMAGE=/vmlinuz-5.15.0-2-amd64 root=UUID=2b82e6c4-442e-4700-a246-3b5c37a722b5 ro tsc=reliable quiet splash slab_common.usercopy_fallback=y intel_iommu=on iommu=pt
[    0.105033] clocksource: hpet: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 133484882848 ns
[    0.125053] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x25629b71302, max_idle_ns: 440795293176 ns
[    0.375011] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.714187] clocksource: Switched to clocksource tsc-early
[    0.730363] clocksource: acpi_pm: mask: 0xffffff max_cycles: 0xffffff, max_idle_ns: 2085701024 ns
[    1.797086] tsc: Refined TSC clocksource calibration: 2593.749 MHz
[    1.797114] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x25632b1b102, max_idle_ns: 440795335950 ns
[    1.797236] clocksource: Switched to clocksource tsc
[  123.639964] tsc: Marking TSC unstable due to KVM discovered backwards TSC
[  123.639981] TSC found unstable after boot, most likely due to broken BIOS. Use 'tsc=unstable'.
[  123.641194] clocksource: Checking clocksource tsc synchronization from CPU 11 to CPUs 0,12,16,18-19,22,30-31.
[  123.641194] clocksource:         CPUs 0,16,18-19,22 ahead of CPU 11 for clocksource tsc.
[  123.641194] clocksource:         CPUs 12,30-31 behind CPU 11 for clocksource tsc.
[  123.641194] clocksource:         CPU 11 check durations 1065ns - 3474ns for clocksource tsc.
[  123.641293] clocksource: Switched to clocksource hpet


# cat /sys/devices/system/clocksource/clocksource0/current_clocksource
hpet
```

**Comment by Raikerian** [comment]
Make sure your cpu has tsc_adjust feature, which seems to be most important one (in my case at least):
```
cat /proc/cpuinfo | grep tsc_adjust
```

Also perhaps some methods from here can help as well https://aws.amazon.com/premiumsupport/knowledge-center/manage-ec2-linux-clock-source/

**Comment by sej7278** [comment]
i seem to only have this lot:

```
tsc
constant_tsc
nonstop_tsc
tsc_deadline_timer
```

**Comment by Raphitpt** [comment]
Well, I tried everything but without success, I think it comes from lenovo from what I could read, so I'm waiting for a fix

**Comment by thenickdude** [comment]
A user on Reddit reported that they had this issue and fixed it by fully shutting down their host and starting it again (warm reboots didn't fix it). Maybe a power cycle is required to resync the TSC between cores.

**Comment by sej7278** [comment]
hmm maybe, i have given up on this given that my gt710 won't even work with monterey without complicated fudging, but i might see if i can give it another go once i've found all my notes!

to me the powercycle thing sounds a bit more like a vendor reset issue with the gfx card (which mine won't have), although i do have a similar problem with a qnap 2.5gbe card - it loses its flash settings or something when you just power off the pc, but pull the cable and it boots again fine afterwards!

**Comment by sickill** [comment]
Some data points from me:

I can install both Big Sur and Monterey without problem _after host cold boot_. After installation they boot fine _when host has just been cold booted_.

After host cold boot dmesg reports `clocksource=tsc` and things are fine. However, after some suspend/resume cycles of the host I can't boot macOS (neither Big Sur nor Monterey). Warm rebooting doesn't help, the host reports `Marking clocksource 'tsc' as unstable` in dmesg and `/sys/devices/system/clocksource/clocksource0/current_clocksource` returns `hpet`.

So it seems a combination of suspend/resume/reboot makes TSC unstable (I do a lot of suspending so not sure yet if just warm reboot without suspending makes it bad as well). Either way, when my system is in this unstable TSC condition I can install Big Sur just fine, although it won't boot after installation. Monterey won't even install.

I tried CPU pinning but no dice.

`tsc_adjust` is not present in `/proc/cpuinfo` for my CPU.

---

Motherboard: ASUS x570 Pro Creator
AGESA: 1.2.0.7
CPU: Ryzen 5700G

**Comment by thenickdude** [comment]
No, that won't fix the problem since the guest will still observe the skewed TSC and panic.

The TSCs need to be resynced on the host

**Comment by sej7278** [comment]
Ah I see, assumed it was just kernel/qemu not passing it to the VM at all

**Comment by sickill** [comment]
Due to other issues I had with my PC I've replaced my Corsair Vengeance LPX DDR4 sticks with G.Skill Ripjaws V ones and suddenly I'm not experiencing any of the above issues anymore. I still see messages about unstable clocksource in dmesg output after warm reboots but they don't seem to affect the ability to boot macOS whatsoever. I can boot and reboot Big Sur now without a problem every single time. So I'll assume the clocksource thing was a red herring in my case.

**Comment by sej7278** [comment]
Yeah bigsur is not a problem for me, just Monterey (or later?)

**Comment by mchrostowski** [comment]
I just successfully resolved this issue on my own system by updating my BIOS.

For me it was indeed a BIOS TSC sync with a lack of TSC Adjust on the affected CPU. I could see my Linux kernel avoiding TSC due to this specifically but macOS would do the non-monotonic panic.

I didn't have to make any further adjustments to my BIOS and I went to stock settings from the downstream project I'm using to launch, Docker-OSX. No extra CPU flags for hypervisor or the like.

**Comment by kszczek** [comment]
Same issue on my Lenovo Legion laptop, although I don't expect to get a BIOS update with the fix - Lenovo seems to focus on the "Linux certified" laptops with their patches. Fortunately there is a workaround but it requires a bit of tinkering. The workaround consists of a set of kernel patches which implement an alternative method of TSC synchronization which fixes this issue on my system. More details in [this Reddit post](https://www.reddit.com/r/Amd/comments/uf0zdf/comment/i6tqak0/).
